### ISSUE TITLE
`comparingOnlyFields` between different classes can fail because of missing fields that aren't being compared

### ISSUE THREAD

**Issue by rd-martin-bonev** [assertj/assertj#2606]
#### Summary
When comparing different classes and specifying `comparingOnlyFields` the assertion could fail if the class `actual` has more fields than the class of `expected` even when those fields were not specified for comparison. The example below compares two objects twice, only reversing which one is `actual` and which one is `expected`. The first assertion passes and the second fails. I'd expect that both pass, or at least that both fail.

Removing `comparingOnlyFields` and setting the same value for `b` similarly results in failure only for the second assertion, but I'm not sure if this is intended or not.

I've tested it with 3.21.0 and 3.22.0.

#### Example

```java
public class ComparingOnlyTest {
  static class A1 {
    public final int a, b, c;
    public A1(int a, int b, int c) {
      this.a = a;
      this.b = b;
      this.c = c;
    }
  }
  static class A2 {
    public final int a, b;
    public A2(int a, int b) {
      this.a = a;
      this.b = b;
    }
  }

  @Test
  public void testComparingOnly() {
    var a1 = new A1(1, 2, 3);
    var a2 = new A2(1, 4);

    Assertions.assertThat(a2).usingRecursiveComparison().comparingOnlyFields("a").isEqualTo(a1);
    Assertions.assertThat(a1).usingRecursiveComparison().comparingOnlyFields("a").isEqualTo(a2);
  }
}
```

**Comment by scordio** [comment]
Duplicate of #2359, fixed by 61ea685ab6fcce01a53d8ebcc0236b0d0c1e04cc.

**Comment by rd-martin-bonev** [comment]
I would say this similar but not a duplicate of #2359 (though it might have been fixed with the same fix). The test in the description passes now with the latest code from the main branch, but I found another issue that might be masking this (#2610).
